### ISSUE TITLE
findByUid and findByIdentifier repository method should not be replaced

### ISSUE THREAD

**Issue by akiessling** [sabbelasichon/typo3-rector#4017]
With #3497 the magic findBy methods are replaced for v12, but they also replace findByUid and probably findByIdentifier calls. Since these methods are directly implemented in the repository class and not deprecated, they should not be replaced https://github.com/TYPO3/typo3/blob/44670dd4df32dad482b01da28a4dd2e0739c0207/typo3/sysext/extbase/Classes/Persistence/Repository.php#L153
They also behave differently and only return a single result, whereas findBy returns multiple results --> this now breaks the migrated code

https://github.com/sabbelasichon/typo3-rector/blob/212545a56cba1dfa1297758cdb32c4b9b086e349/src/Rector/v12/v3/typo3/MigrateMagicRepositoryMethodsRector.php#L74

**Comment by akiessling** [comment]
Thinking further about it: should that check have taken care of it, so that existing methods won't be modified? 

```
if ($methodReflection instanceof MethodReflection) {
```

**Comment by sabbelasichon** [comment]
@akiessling Do you have saschaegerer/phpstan-typo3 installed?

**Comment by akiessling** [comment]
Good guess, it is installed but wasn't the problem. 

After some debugging i found the issue now..

The old code was 
```
/**
 * contactRepository
 *
 * @var \My\Ext\Domain\Repository\ContactPersonRepository
 * @TYPO3\CMS\Extbase\Annotation\Inject
 */
protected $contactPersonRepository = null;
```

Thus i migrated this to `protected ?ContactPersonRepository $contactPersonRepository = null;`  since phpstan wanted strict types etc.
But if it is nullable, then `$methodReflection = $this->reflectionResolver->resolveMethodReflectionFromMethodCall($node);` fails to find the method.

I'm not sure as to why the old code had `= null`

So to sum it up

This works
```
    /**
     * @var ContactPersonRepository
     */
    protected $contactPersonRepository = null;
```

this also works
```
protected ContactPersonRepository $contactPersonRepository;
```

This doesn't

```
protected ?ContactPersonRepository $contactPersonRepository = null;
```

I'm not sure if this is something that the MigrateMagicRepositoryMethodsRector should try to fix or work around, but maybe we could/should add an explicit handling of findByUid and findByIdentifier since these methods are guaranteed to exist in \TYPO3\CMS\Extbase\Persistence\Repository?

**Comment by sabbelasichon** [comment]
@akiessling Thanks for your investigation. We should still try to look into the concrete class. This should be fixable.

**Comment by akiessling** [comment]
Hey @sabbelasichon 

Thanks for taking care!

I updated my project to dev-main (36c7bf3) and readded the nullable config in the controller for testing.

```
protected ?ContactPersonRepository $contactPersonRepository = null;
```

--> the rewrite of the existing method in the controller no longer happens, my call of the findByUids method is kept  ✅ 


But in the repository class, this method just loops over the given uids and calls findByUid

```
class ContactPersonRepository extends Repository
{
    /**
     * @return ContactPerson[]
     */
    public function findByUids(array $uids): array
    {
        $result = [];
        foreach ($uids as $uid) {
            /** @var ContactPerson $contact */
            $contact = $this->findByUid($uid);
            if ($contact) {
                $result[] = $contact;
            }
        }
        return $result;
    }
}
```

This is now wrongly rewritten to the `findBy` call

```
    ---------- begin diff ----------
@@ @@
         $result = [];
         foreach ($uids as $uid) {
             /** @var ContactPerson $contact */
-            $contact = $this->findByUid($uid);
+            $contact = $this->findBy(['uid' => $uid]);
             if ($contact) {
                 $result[] = $contact;
             }
    ----------- end diff -----------
```

**Comment by sabbelasichon** [comment]
@akiessling Good catch. Fixed it

**Comment by akiessling** [comment]
Updated, tested --> looks good! Thank you @sabbelasichon !

**Comment by simonschaufi** [comment]
@akiessling by the way, your SQL can be improved by requesting all records at once and them loop over them to create the objects. If they are only 5 records, it doesn't make a big difference but if you request a couple of 100 records, you would have just one db request instead of 100.

**Comment by akiessling** [comment]
Yeah, i know. But AFAIK we still can't use ORDER BY FIELD and you need to take care of the ordering yourself afterwards. Thus for a small plugin selection it's often not worth the boilerplate code
